### DISCUSSION TITLE
hate to borther but I got a very complicated usage

### DISCUSSION THREAD

**Discussion by luohao123** [simolus3/drift#2428] [Q&A]
Hi, I have a table defined like this:

```dart
@DataClassName("DbMessage")
class Messages extends Table {
  TextColumn get id => text()();
  TextColumn get type => text()();
  TextColumn get fromId => text().named("from_id")();
  TextColumn get fromName => text().named("from_name").nullable()();
  // TextColumn get fromAvatar => text().named("from_avatar").nullable()();
  TextColumn get toId => text().named("to_id").nullable()();
  TextColumn get toName => text().named("to_name").nullable()();
  TextColumn get textClm => text().named("text")();
  TextColumn get roomId => text().named("room_id")();
  TextColumn get originality => text()();
  TextColumn get attachment => text().nullable()();
  TextColumn get thumbnail => text().nullable()();
  TextColumn get originalId => text().named("original_id").nullable()();
  IntColumn get sendTime => integer().named("send_time")();
  IntColumn get size => integer().withDefault(const Constant(0)).nullable()();
  TextColumn get status => text().withDefault(const Constant("pending"))();
  TextColumn get mime => text().nullable()();
  RealColumn get longitude => real().nullable()();
  RealColumn get latitude => real().nullable()();

  @override
  Set<Column> get primaryKey => {id};
}

```

Now, I want query all messages  belongs to a contact out along with every message's replied child messages.

How do i retrive it?

```dart
Stream<List<DbMessageWithDbContact>> getMessagesWithContactByRoomId(
      String roomId) {
    final query = select(messages)
        .join([leftOuterJoin(contacts, contacts.id.equalsExp(messages.fromId))])
      ..where(messages.roomId.equals(roomId))
      ..orderBy([
        OrderingTerm(expression: messages.sendTime, mode: OrderingMode.desc)
      ]);
    return query.watch().map((rows) {
      return rows.map((row) {
        return DbMessageWithDbContact(
          row.readTable(messages),
          row.readTableOrNull(contacts),
        );
      }).toList();
    });
  }
```

this is currently I got, I wanna query out all messages's replied messages as well.

I am new to sql, chatgpt gives me this 
```sql
SELECT m1.*, m2.*
FROM DbMessage AS m1
LEFT JOIN DbMessage AS m2 ON m1.id = m2.original_id
```
how can i convert it to drift syntax?

**Comment by simolus3** [comment]
> every message's replied child messages.

Generally, SQL doesn't have a concept of lists in the results. In pure SQL, you can't write a query that directly returns something like this:

```dart
class MessageWithChildren {
  final DbMessage message;
  final List<DbMessage> children;
}
```

However, you could write a query that returns messages in an order that makes it trivial to create this structure in Dart. Let's say that these messages exist:

- Message 1 without a parent.
- Message 2 without a parent.
- Message 3 in reply to 1.
- Message 4 without a parent.
- Message 5 in reply to 3.
- Message 6 in reply to 1.

I assume that you want this structure in Dart, right?

```dart
[
   MessageWithChildren(message1, [
     MessageWithChildren(message3, [
       MessageWithChildren(message5, []),
     ]),
     MessageWithChildren(message6, []),
   ]),
   MessageWithChildren(message2, []),
   MessageWithChildren(message4, []),
]
```

For this, you can write a SQL query that recursively loads nested messages into a flat result, thus returning all messages in the right order (1, 3, 5, 6, 2, 4). By later looking up the parent id in Dart (knowing that results come in depth-first order), you can reconstruct a tree of messages.

```sql
WITH RECURSIVE
  source AS (SELECT * FROM messages WHERE room_id = :room), -- All messages we're interested in
  topLevel AS (SELECT * FROM source WHERE original_id IS NULL), -- Messages at the root of the conversation
  nested AS (
    SELECT *, 0 AS depth FROM topLevel, -- Load top-level messages first
    UNION ALL -- And then for each found message
    SELECT 
      children.*, depth + 1
    FROM messages AS children, nested
    WHERE children.original_id = nested.id -- find child messages
    ORDER BY depth DESC, send_time DESC -- depth-first search, sort by send time for messages on the same level
  )
SELECT * FROM nested;
```

Before you copy this, I strongly recommend you do some research on SQL and [WITH clauses](https://sqlite.org/lang_with.html) in particular - you're asking for something that is very complicated to do in a single SQL query.

You could then run this query as SQL with [`customSelect`](https://drift.simonbinder.eu/api/drift/databaseconnectionuser/customselect) and iterate over results to build the tree structure. If you need the sender for each message, you could add a join to the final query.

**Reply by luohao123** [reply]
Thank u so much for dedicated answer. 

The SQL looks like pretty complicated for me. Actually my situation is a little bit simpler: one message should only consider only one parent reply to. In otherwords, every replied msg **should only have one single parent**, (but one message could have multiple replies).

I thought is that possible by chaning a little bit with my existed code:

```dart
Stream<List<DbMessageWithDbContact>> getMessagesWithContactByRoomId(
      String roomId) {
    final query = select(messages)
        .join([leftOuterJoin(contacts, contacts.id.equalsExp(messages.fromId))])
      ..where(messages.roomId.equals(roomId))
      ..orderBy([
        OrderingTerm(expression: messages.sendTime, mode: OrderingMode.desc)
      ]);
    return query.watch().map((rows) {
      return rows.map((row) {
        return DbMessageWithDbContact(
          row.readTable(messages),
          row.readTableOrNull(contacts),
        );
      }).toList();
    });
  }
```

Since I might need want keep the exited struct on changed (with Contacts queried out as well).

**Reply by simolus3** [reply]
> (but one message could have multiple replies)

Is it possible to nest them? E.g, could a message that was a reply have replies as well?

I still don't understand the exact query you're interested in: Basically, you want to look up all messages in a conversation and load the sending contact for each message? How do replies come into play here? If you have an example table of messages and the results you need, perhaps I understand your query better and can come up with a solution.

**Reply by luohao123** [reply]
@simolus3 
> Is it possible to nest them?

I don't consider this siutation, one message have 1 or 1+replies, but replies didn't have replies.

Let me describe it simple, basically, i have this query out Messages out right? (although along with contacts), but we consider Mesages only:

```dart
Stream<List<DbMessageWithDbContact>> getMessagesWithContactByRoomId(
      String roomId) {
    final query = select(messages)
        .join([leftOuterJoin(contacts, contacts.id.equalsExp(messages.fromId))])
      ..where(messages.roomId.equals(roomId))
      ..orderBy([
        OrderingTerm(expression: messages.sendTime, mode: OrderingMode.desc)
      ]);
    return query.watch().map((rows) {
      return rows.map((row) {
        return DbMessageWithDbContact(
          row.readTable(messages),
          row.readTableOrNull(contacts),
        );
      }).toList();
    });
  }
```

Now the messages defines like this:

```dart
@DataClassName("DbMessage")
class Messages extends Table {
  TextColumn get id => text()();
  TextColumn get type => text()();
  TextColumn get fromId => text().named("from_id")();
  TextColumn get fromName => text().named("from_name").nullable()();
  // TextColumn get fromAvatar => text().named("from_avatar").nullable()();
  TextColumn get toId => text().named("to_id").nullable()();
  TextColumn get toName => text().named("to_name").nullable()();
  TextColumn get textClm => text().named("text")();
  TextColumn get roomId => text().named("room_id")();
  TextColumn get originality => text()();
  TextColumn get attachment => text().nullable()();
  TextColumn get thumbnail => text().nullable()();
  TextColumn get originalId => text().named("original_id").nullable()();
}
```

Is clear right. Now I want the Messages's `originalId` attached into it point to. For example, I have:

Message0
```
{'id': 0, text: "hi", original_id: null}
```

Message1
```
{'id': 1, text: "hi", original_id: 0, originlity: REPLY}
```

Then I need queryout Message0 like this:

```
{'id': 0, text: "hi", original_id: null, replies: [{'id': 1, text: "hi", original_id: 0, originlity: REPLY}] }
```

In short, how to get the child if a column same with queried id.
